Show only allowed widgets for a certain main parameter

imenattatra opened the pull request:
## What type of PR is this? 

This is an extension of an existing feature . The existing feature shows all the dashboard's widgets in the dashboard page, no matter what the dashboard parameters values are. 

In our extension (the code in this PR) we add the possibility to only show the allowed widgets according to a parameter value.

You can see this happening on staging [in this video ](https://drive.google.com/file/d/17ppzCbC1rztZ6bR25BrldsddPTbkqTdh/view?usp=sharing)

![hideandsee](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExMDVmMTh4cmxxNHhsNzN3NjhoeDE0MHJmeGhmMmM2b29reXZrMGVpOSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/MzZWQV6VGPMvTkaxFc/giphy.gif)

## Pre-Description

- This has been developed : 
Assuming that the dashboard parameters that we use -in our case- never share common values. 


## Description

After we include this code in our image of redash, we will be able to only show the widgets that we want for the dashboards that we  wish to customise.

The introduced code is a transition before the final state where we will be getting the list of allowed_widgets from our product re****_mo**** directly.

But, in order to make it work for now, we need to have certain things set: 

1- The **allowed_widgets_dashboardID** query 

* The alllowed_widgets query needs to have its name in the format `allowed_widgets_{dashboardID} `
* If a dashboard has no related alllowed_widgets query, it will simply show all the widgets without filtering.
*The query should have the following columns : `main_parameter` and `widgets`, where `main_parameter` will include the parameter values , for each parameter value that we want to customise, we specify the list of widgets we want to show, a widget is identified by its visualisation name

example : 

```sql 

select "collector123" as "main_parameter",  "['heat1', 'water2']" as "widgets"
union select "zone456" as "main_parameter",  "['heat1', 'heat2']" as "widgets"

```
2- The visualisations names 

* For every widget that we want to show when we want, we should add a name to its visualisation (so far only one ... in the futur many)

example : `'heat1'` as visualization name

## How is this tested?

- [x] Unit tests (pytest, jest)
- [x] Manually , check video mentioned above

## What is next?

![](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExcnp4Y3YxNGV4a25kYzFsMHowMTZ3NjkxdGV6bTFubDlrdDZnNWR0dyZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/9CrVjsT4y6Jxu/giphy.gif)

- using the widgets tags instead of the visualisation names (but will they really be better than the visualisation names)
- build the allowed_widgets query directly from our product re****_mo****
- discuss the positioning of the widgets if needed


## Related Tickets & Documents

Closes https://github.com/orgs/metr-systems/projects/18/views/1?pane=issue&itemId=74429589
Closes https://github.com/orgs/metr-systems/projects/18/views/1?pane=issue&itemId=74429568
Closes https://github.com/orgs/metr-systems/projects/18/views/1?pane=issue&itemId=74429622

imenattatra committed:
add code to show only allowed widgets per parameter controller_param

From the backend side:
- the backend gets the list of allowed-widgets for every controller by reading the query allowed_widgets that is supposed to exist for every organization
- if this restriction does not exist we simply do not add allowed_widgets to the response

From the frontend side:
- we get the controller param value, this param would be part of any query (every widget has a query), this asumption needs to be confirmed
-we get the allowed widgets for that controller from the response we receive from backend
- we filter widgets before preparing them
- if an error happens during this process, we just return all dashboad widgetsmake sure we return the data keys we want and a correct timestamp format for the transmission time

imenattatra committed:
filter for allowed widgets on the level of the ´useDashboard´ hook

It seems more correct to have this filter for allowed widgets when rendering the dashboard to make sure it works correctly even if the param changes, not only that but also here we can rely on ´dashboard.getParametersDefs();´ which is much better than relying on the param from the query like done in the previous commit

imenattatra committed:
only filter ´allowed_widgets´ query by name

since it will be one query shared by all organizations

imenattatra committed:
reset dashboard to have all the widgets before filtering the widgets per param

imenattatra committed:
use generic parameter name and specify the dashboard id in the allowed widgets query

imenattatra committed:
filter widgets by first parameter value that has an entry in the allowed_widgets dictionary

imenattatra committed:
add command to run one test to the readme

imenattatra committed:
update  to make it easy to test

imenattatra committed:
extract  function to a separated file

imenattatra committed:
update 'getAllowedWidgetsForCurrentParam' and introduce a unit test for it

- make it use general names like paramValue and allowedWidgets
- make sure we compare against the widget visualization name
- make sure we only attempt to filter if dashboardAllowedWidgets is defined

imenattatra committed:
move 'get_allowed_widgets_info' out of 'DashboardResource' and introduce unit tests for it

the move was in order to have completely separated from 'DashboardResource' and only taking the 'dashboard_id' as parameter

imenattatra committed:
keep minimum of info for the parameters to test

imenattatra committed:
add comments to code for our futur-selves

imenattatra committed:
use 'widgets' instead of 'allowed_widgets' as querycolumn name as specified in the related issue

charinass commented on client/app/pages/dashboards/hooks/utils.test.js:
whew! js test! i'm so surprised 😲 beautiful 😻

thiagogds replied:
They're really cool and simple too!

imenattatra replied:
thanks :D

charinass reviewed (approved):
Awesome PR 😻 thank you for the explanation and loved that you put tests for js.

thiagogds commented on client/app/pages/dashboards/hooks/utils.js:
```diff
@@ -0,0 +1,18 @@
+      if (dashboardAllowedWidgets !== undefined && dashboardAllowedWidgets.hasOwnProperty(paramValue)) {
+        // filter the widgets to process
+        const allowedWidgets = dashboardAllowedWidgets[paramValue];
+        return dashboardAllWidgets.filter(widget => allowedWidgets.includes(widget.visualization.name));
```
You're missing the case where the visualization name has more than 1 widget. For example: "heat1;heat2"

imenattatra replied:
according to what i got so far, one visualisation is really linked to one query -> to one widget ... but then we can add many visualisations to one widget and give each one a name, in this case our list of allowed widgets can have all of them or some of them or none.. as we want... the current code supports this case. 

It seems that in order to add many visualisations to a dashboard, we add every time a widget and then pick the visualisation that we want ... this is why it is already supported

am I missing something?

thiagogds replied:
My point isn't related to Redash, but to remote monitoring. @helenalebreton Can you help to see if this makes sense?

I was thinking that the same visualization could be used in a case of more than 1 group. For example:

We would have a visualization named "heat1;heat2". This is related to the same query, where just the controller ID changes.

Then imagine we have 2 controllers. One in a building with just just one heat circuit, so it would have a group type Heat and number 1. And another building with 2 heating circuits, then it would have two groups with type Heat, one with number 1 and another with number 2.

My expectation then is that the allowed_widgets query would have this data:

|main_parameter|widgets|
|--------|--------|
|123| ['heat1'] |
|456| ['heat1', 'heat2']|

Then, for me, that visualization should work on both controllers. But now that I'm writing it I'm not sure anymore that this is a valid use case. 

Right now we need to have a visualization for heat1 and another visualization for heat2. And both would show for controller 456

imenattatra replied:
thank you for your explanation , after @helenalebreton [confirmation](https://github.com/metr-systems/redash/pull/3#issuecomment-2312434216) , i updated the code to support a compound visrtualisation [name in this commit](https://github.com/metr-systems/redash/pull/3/commits/38afa1c89eb2276a4c8823adbcef7d8321a60c2e)

thiagogds commented on client/app/pages/dashboards/hooks/useDashboard.js:
```diff
@@ -130,6 +130,14 @@ function useDashboard(dashboardData) {
   const loadDashboard = useCallback(
     (forceRefresh = false, updatedParameters = []) => {
+      // filter widgets to show from all the widgets according to the current parameters
+      dashboardRef.current.widgets = dashboardRef.current.saved_all_widgets;
+      dashboardRef.current.widgets = getAllowedWidgetsForCurrentParam(
+        dashboardRef.current.getParametersDefs(),
+        dashboardRef.current.allowed_widgets,
+        dashboardRef.current.widgets
+      );
```
Do we need to first set the `dashboardRef.current.widgets = dashboardRef.current.saved_all_widgets;` and then do it again? Can't we do:

```suggestion
      dashboardRef.current.widgets = getAllowedWidgetsForCurrentParam(
        dashboardRef.current.getParametersDefs(),
        dashboardRef.current.allowed_widgets,
        dashboardRef.current.saved_all_widgets
      );
```

imenattatra replied:
thanks, i updated it in [the latest commit ](https://github.com/metr-systems/redash/pull/3/commits/625fe23e8df46b392c8e47e63e335e85996180ca)

thiagogds reviewed (changes requested):
Very cool! 

Do you need help to deploy it staging? I'm super curious to see it live hehe

imenattatra commented:
> Very cool!
> 
> Do you need help to deploy it staging? I'm super curious to see it live hehe

it is already in [staging](https://dashboard.staging.metr.systems/staging/dashboards/1-test?p_parameter=foo) :D

berinhard commented on redash/settings/__init__.py:
```diff
@@ -16,6 +17,8 @@
 from redash.settings.organization import DATE_FORMAT, TIME_FORMAT  # noqa
 
+load_dotenv()
+
```
Even though this is definitely better than using `os.environ.get` directly, I'd still keep on with the standards that Redash has already been using. My arguments behind that are:

1. ["There should be one-- and preferably only one --obvious way to do it."](https://peps.python.org/pep-0020/)
2. It'll become harder to open PRs and propose changes to Redash's main if we change some part of the project's strategies;
3. For the same reason, it can become harder to merge in changes from Redash's main;

imenattatra replied:
Thanks a lot for your feedback on this @berinhard . I had to rebase my source branch here to be updated with metr-main because this update regarding the usage of `dotenv` has been already merged there in an earlier PR that you can find [here](https://github.com/metr-systems/redash/pull/2/files) . As far as i remember , the project was not uploading the needed en variable before adding `load_dotenv()`, do you want to maybe pair on it and we can look at it together?

berinhard commented on redash/handlers/dashboards.py:
```diff
@@ -194,6 +213,11 @@ def get(self, dashboard_id=None):
         self.record_event({"action": "view", "object_id": dashboard.id, "object_type": "dashboard"})
 
+        # add allowed_widgets to the dashboard information to return in case it exists and it is not empty
+        allowed_widgets = get_allowed_widgets_info(dashboard_id)
+        if allowed_widgets:
+            response["allowed_widgets"] = allowed_widgets
```
I'm not suggesting a change here, but I just wanna to talk more about strategies to have less intrusive changes on Redash. We can use the idea of merge conflicts as an heuristic of how "intrusive" our changes are. Here, for example, there are higher chances of merge conflicts since we're changing the view code itself. So, the question is: how could we avoid this?

I feel a good candidate here could be a Python decorator. By using it, we could keep the view code exactly the way it is and only change its signature to include a new decorator. This significantly reduces the change of, not only conflicts from happening, but also unexpected bugs. The decorator would decorate this view an update its response the same way it's doing here, but in a more isolated way;

For example, let's imagine that the Redash's main introduced changes to this view and, during the conflict resolution, for any reason we've done it wrongly and erased one of the new lines from Redash's main. This will create a bug and we'll be responsible to fix this.

But, again, no need to change anything. I'm only bringing up this conversation so we can consider such topics within newer PRs.

imenattatra replied:
thanks @berinhard I liked the decorator idea and tried its implementation [here](https://github.com/metr-systems/redash/pull/3/commits/02f9bc206383881d0c9b7a178e0612f25411fa0d)

berinhard commented on redash/handlers/dashboards.py:
```diff
@@ -135,6 +135,25 @@ def get(self):
+    if query:
+        data = query.latest_query_data.data["rows"]
+        for row in data:
+            if "main_parameter" in row.keys() and "allowed_widgets" in row.keys():
+                allowed_widgets[row["main_parameter"]] = row["allowed_widgets"]
```
Can we add these `main_prameter` and `allowed_widgets` as variables or constants definition in this method?

I feel that there's a certain level of [data coupling](https://en.wikipedia.org/wiki/Coupling_(computer_programming)) going on here, so it'd be nice to reduce the chance of misreading information.

imenattatra replied:
thanks @berinhard please find update for it here

berinhard commented on metr_docs/README.md:
```diff
@@ -195,9 +197,25 @@ Enabling Pre-commit check before commit.
+- command to run one frontend test named "testName" is `yarn jest -t "testName"`
+  checking your installed dependencies for any failing tests
+
+- for installing cypress on mac
```
Thanks for adding specific information to our custom readme file :+1:

imenattatra replied:
thanks :D

berinhard reviewed (changes requested):
I focused on reviewing more the Python part since I've saw that @thiagogds already left a pair of comments in the JS part, that I don't know how to review properly.

Anyway, thanks **A LOT** for adding JS tests. I feel way more comfortable about merging this since you've added them. Kudos!

The only thing that I really feel as a blocker is the change on `settings` to use the `dotenv`  project. Besides that, this looks great and promissing!

imenattatra committed:
remove not needed line

use 'saved_all_widgets' directly without passing them to 'widgets'

Co-authored-by: thiagogds <thiagogds14@gmail.com>

imenattatra committed:
Merge branch 'show-only-allowed-widgets-for-a-certain-main-parameter' of https://github.com/metr-systems/redash into show-only-allowed-widgets-for-a-certain-main-parameter

helenalebreton commented:
@imenattatra so we do need the possibility for several widget tags in order to test properly because I would like to test with GAG. Since we have no UI yet for the tags, any way we can simulate several tags using the visualisation name is fine with me.

imenattatra committed:
make sure textbox widgets are not filtered

textboxes do not have a visualization and we don't want them to be affected by the filtering

thiagogds commented on client/app/pages/dashboards/hooks/utils.js:
```diff
@@ -0,0 +1,20 @@
+export function getAllowedWidgetsForCurrentParam(dashboardParameters, dashboardAllowedWidgets, dashboardAllWidgets) {
+  try {
+    // filter the widgets by keeping only the ones we want
+    for (var param of dashboardParameters) {
+      const paramValue = param.value;
```
I'm not sure if it's a good idea to filter using any parameter. This can create some inconsistencies.  I prefer to have a hard-coded name for the parameter we will use.

What do you think @helenalebreton?

Right now if a parameter value matches any row in the allowed_widget it will be used.

thiagogds replied:
Actually, it will match on the first parameter that has a known value. I think this logic is too implicit.

helenalebreton replied:
@thiagogds do you mean the main parameter ? I am not sure I follow - maybe a quick call on that one.

helenalebreton replied:
or do you mean the fact that if one value from the widget field in the allowed_widgets query matches at least one tag of a widget (currently simulated with visualisation name), then we show the widget ?

thiagogds commented on client/app/pages/dashboards/hooks/utils.test.js:
```diff
@@ -0,0 +1,131 @@
+    ]).toStrictEqual(result);
+  });
+
+  test("filter correctly in case of many parameters", () => {
```
Related to my comment about parameters. This isn't clear that the expected behaviour is to match the first parameter. If we keep this, we should rename the test to "filter correctly using the first matching parameter of many"

imenattatra committed:
add support to a compound visualisation name

helenalebreton commented:
with
prep_groups as (
SELECT device_id , concat('''',type,number,'''') as visu_list_prep
FROM core_group
where device_id in (13, 69, 155) and device_type_id = 12),
data_prep as (
SELECT device_id , STRING_AGG(visu_list_prep, ',') as visu_list
FROM prep_groups
GROUP BY device_id)
select device_id as main_parameter, concat('[',visu_list,']') as widgets
from data_prep

helenalebreton commented:
and something like that when the dashboard is based on controller but you also have other devices like for GAG:

with

prep_groups_controller as (
SELECT distinct device_id , --concat('''',type,number,'''') as visu_list_prep, 
building_uuid
FROM core_group
where device_id in (13, 69, 155) and device_type_id = 12),

prep_groups as (
select c.building_uuid, concat('''',c.type,c.number,'''') as visu_list_prep 
from core_group c
inner join prep_groups_controller p on c.building_uuid=p.building_uuid
where p.device_id in (select device_id from prep_groups_controller)
group by c.building_uuid, c.type, c.number
),

data_prep as (
SELECT building_uuid , STRING_AGG(visu_list_prep, ',') as visu_list
FROM prep_groups
GROUP BY building_uuid)

select c.device_id as main_parameter, concat('[',visu_list,']') as widgets
from data_prep p
left join prep_groups_controller c on c.building_uuid=p.building_uuid

helenalebreton commented:
it will get y bit trickier with OMS groups because they have the same types and number as other groups linked to bus devices but they would be different widgets. For Example refley may be monitored like a controller or with an OMS sensor. in both cases I would have a reflex1 group BUT one will have a device linked to it and the other one not. 
Therefore I think we can build the list checking this as well and provided different widgets names fro bus or OMS devices.

I have not written this query but I believe it is doable. I will double check with Andreas and Chris on monday

helenalebreton commented:
@imenattatra, @thiagogds explained me his comments on the parameter value. Now I understand we are there speaking about the possibility to have 2 'main_parameters' to filter the dashobard and that the value of both currently would be consiedered by the allowed_widget query. I agree with @thiagogds that we can and should restrict to only the value of 1 parameter being considered. We will have only 1 main parameter and can 'hard code' the UI dynamic on that.

helenalebreton commented:
> it will get y bit trickier with OMS groups because they have the same types and number as other groups linked to bus devices but they would be different widgets. For Example refley may be monitored like a controller or with an OMS sensor. in both cases I would have a reflex1 group BUT one will have a device linked to it and the other one not. Therefore I think we can build the list checking this as well and provided different widgets names fro bus or OMS devices.
> 
> I have not written this query but I believe it is doable. I will double check with Andreas and Chris on monday

@imenattatra I am pretty sure we will find a way for this. You can continue on the milestone

imenattatra committed:
extract main_parameter and widgets list column names

imenattatra committed:
extract the code that adds the allowed widgets information to a decorator

imenattatra committed:
correct command in documentation
